File: node/src/inventoryItemEventSourcing.js
const { Created, Stocked, Sold, Activated, Deactivated } = require('./inventoryEvent');
const { Create, Stock, Sell, Activate, Deactivate } = require('./inventoryCommand');
const { InventoryItem } = require('./inventoryItem');

class InventoryItemEventSourcing {
    static Init() {
        return new InventoryItem(null, null, 0, false);
    }

    static Apply(event, state) {
        if (!event) return state;
        switch (event.constructor) {
            case Created:
                return new InventoryItem(event.inventoryId, event.name, state.count, state.active);
            case Stocked:
                return new InventoryItem(state.id, state.name, state.count + event.count, state.active);
            case Sold:
                return new InventoryItem(state.id, state.name, state.count - event.count, state.active);
            case Deactivated:
                return new InventoryItem(state.id, state.name, state.count, false);
            case Activated:
                return new InventoryItem(state.id, state.name, state.count, true);
            default:
                return state;
        }
    }

    static Execute(command, state) {
        switch (command.constructor) {
            case Create:
                if (state.id === null && state.name === null) {
                    return new Created(command.inventoryId, command.name);
                }
                break;
            case Stock:
                if (state.id !== null && command.count > 0) {
                    return new Stocked(command.count);
                }
                break;
            case Sell:
                if (state.id !== null /*&& state.count - command.count >= 0 && state.active && command.count > 0*/) {
                    return new Sold(command.count);
                }
                break;
            case Deactivate:
                if (state.id !== null && state.active) {
                    return new Deactivated();
                }
                break;
            case Activate:
                if (state.id !== null && !state.active) {
                    return new Activated();
                }
                break;
        }
        return null; // return new Errored(command, state);
    }
}

module.exports = { InventoryItemEventSourcing };

File: dotnet/csharp/src/InventoryItemEventSourcing.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace EventSourcing
{
    public class InventoryItemEventSourcing
    {
        public static InventoryItem Init => new InventoryItem(Guid.Empty, null, 0, false);

        public static InventoryItem Apply(InventoryEvent @event, InventoryItem state)
        {
            switch (@event)
            {
                case InventoryEvent.Created created:
                    return new InventoryItem(created.InventoryId, created.Name, state.Count, state.Active);
                case InventoryEvent.Stocked stocked:
                    return new InventoryItem(state.Id, state.Name, state.Count + stocked.Count, state.Active);
                case InventoryEvent.Sold sold:
                    return new InventoryItem(state.Id, state.Name, state.Count - sold.Count, state.Active);
                case InventoryEvent.Deactivated _:
                    return new InventoryItem(state.Id, state.Name, state.Count, false);
                case InventoryEvent.Activated _:
                    return new InventoryItem(state.Id, state.Name, state.Count, true);
                default:
                    return state;
            }
        }

        public static InventoryEvent Execute(InventoryCommand command, InventoryItem state)
        {
            switch (command)
            {
                case InventoryCommand.Create create when state.Id == Guid.Empty && state.Name == null:
                    return new InventoryEvent.Created(create.InventoryId, create.Name);
                case InventoryCommand.Stock stock when state.Id != Guid.Empty && stock.Count > 0:
                    return new InventoryEvent.Stocked(stock.Count);
                case InventoryCommand.Sell sell when state.Id != Guid.Empty /*&& state.Count - sell.Count >= 0 && state.Active && sell.Count > 0*/:
                    return new InventoryEvent.Sold(sell.Count);
                case InventoryCommand.Deactivate _ when state.Id != Guid.Empty && state.Active:
                    return new InventoryEvent.Deactivated();
                case InventoryCommand.Activate _ when state.Id != Guid.Empty && !state.Active:
                    return new InventoryEvent.Activated();
                default:
                    return null; // return new InventoryEvent.Errored(command, state);
            }
        }
    }
}
